### PR TITLE
Unwrap HttpRequest and HttpResponse in Session-based CSRF strategy

### PR DESCRIPTION
**Pull request by erdlet**
see: #151
Signed-off-by: Erdle, Tobias <tobias.erdle@innoq.com>

### CODE REVIEW

File: core/src/main/java/org/eclipse/krazo/core/HttpCommunicationUnwrapper.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    /**
+     * Unwraps a {@link javax.servlet.http.HttpServletRequest} from a custom wrapper into its original class.
+     *
+     * @param obj the object to unwrap
+     * @param type the target type of the object
+     * @param <T> the generic type of the object
+     * @return the unwrapped {@link javax.servlet.http.HttpServletRequest}
+     */
+    <T extends HttpServletRequest> T unwrapRequest(final T obj, final Class<T> type);
```
Comment (chkal) [inline core/src/main/java/org/eclipse/krazo/core/HttpCommunicationUnwrapper.java]: I'm wondering if it makes sense to simplify this to:

    HttpServletRequest unwrapRequest(HttpServletRequest request);

Do we need the generics here?

File: core/src/main/java/org/eclipse/krazo/core/HttpCommunicationUnwrapper.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Defines an SPI which allows the client to unwrap a HTTP request or response
+ * into a standard {@link javax.servlet.http.HttpServletRequest} or {@link javax.servlet.http.HttpServletResponse}.
+ * <p>
+ * This SPI is intended to be used in case a JAX-RS implementation uses its own wrapper classes for HTTP requests and responses.
+ */
+public interface HttpCommunicationUnwrapper {
```
Comment (chkal) [inline core/src/main/java/org/eclipse/krazo/core/HttpCommunicationUnwrapper.java]: Not 100% sure I like the name. But naming is hard. :-)

Maybe something like `ServletUnwrapper`, `ServletClassUnwrapper`? Not sure if this is better though. :-)

File: cxf/src/main/java/org/eclipse/krazo/cxf/http/CXFHttpCommunicationUnwrapper.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.lang.reflect.InvocationTargetException;
+
+/**
+ * Unwrapper for CXF which extracts the original {@link HttpServletRequest} and / or {@link HttpServletResponse}
+ * from the wrapper which are used e.g. by Liberty. This is required because the wrappers don't use the official wrapper base classes
+ * and therefore Liberty fails to forward such requests because unwrapping isn't possible.
+ */
+public class CXFHttpCommunicationUnwrapper implements HttpCommunicationUnwrapper {
```
Comment (chkal) [inline cxf/src/main/java/org/eclipse/krazo/cxf/http/CXFHttpCommunicationUnwrapper.java]: Don't we need `@Priority` to ensure correct ordering?

File: core/src/main/java/org/eclipse/krazo/core/HttpCommunicationUnwrapper.java (hunk 1 above)
Comment (erdlet) [reply core/src/main/java/org/eclipse/krazo/core/HttpCommunicationUnwrapper.java]: I'm not sure, maybe we can do this explicit. I'll check this

File: core/src/main/java/org/eclipse/krazo/core/HttpCommunicationUnwrapper.java (hunk 2 above)
Comment (erdlet) [reply core/src/main/java/org/eclipse/krazo/core/HttpCommunicationUnwrapper.java]: Hmmm I'm struggling with this naming too, but I think `Servlet...` indicates, that we work with servlets and not HTTP requests and responses :thinking:

File: cxf/src/main/java/org/eclipse/krazo/cxf/http/CXFHttpCommunicationUnwrapper.java (hunk 3 above)
Comment (erdlet) [reply cxf/src/main/java/org/eclipse/krazo/cxf/http/CXFHttpCommunicationUnwrapper.java]: Yes we do, I just forgot it (was a long day yesterday :) )

**Comment by erdlet** [comment]
Hmmm, this seems to fail in Liberty for some reason. I'll have a look on my machine what causes this

**Comment by chkal** [comment]
> Hmmm, this seems to fail in Liberty for some reason. I'll have a look on my machine what causes this

But Travis is green! Or am I missing something?

**Comment by erdlet** [comment]
The number of failed tests increased within this PR because of some server error. I need to check this on my local machine to be able to see the server logs

**Comment by erdlet** [comment]
Seems that most of the test failures are caused by #157. I opened #161 to run all testsuite tests with resource autodiscovery, so we know if this change is causing new trouble or not.

**Comment by erdlet** [comment]
I tested the problem from #151 again with the fixed version and the failure didn't occur again.

Anyway, I recognized that the test for the CSRF HTTP Header fails in the testsuite for Liberty, because the header is missing. But when running e.g. the test application from #151, everything works and the header is present in the HTTP response when loading the page. Any idea, what this could be?

File: core/src/main/java/org/eclipse/krazo/core/HttpCommunicationUnwrapper.java (hunk 1 above)
Comment (erdlet) [reply core/src/main/java/org/eclipse/krazo/core/HttpCommunicationUnwrapper.java]: I removed the generics here, works fine :+1:

File: core/src/main/java/org/eclipse/krazo/core/HttpCommunicationUnwrapper.java (hunk 2 above)
Comment (chkal) [reply core/src/main/java/org/eclipse/krazo/core/HttpCommunicationUnwrapper.java]: I agree. While the MVC spec doesn't directly depend on Servlet, Krazo only works in Servlet environments at the moment. Therefore, using Servlet in the name may be ok. However, I'm not strong on this one. So let's keep it as it is.

**Comment by chkal** [comment]
> Anyway, I recognized that the test for the CSRF HTTP Header fails in the testsuite for Liberty, because the header is missing. But when running e.g. the test application from #151, everything works and the header is present in the HTTP response when loading the page. Any idea, what this could be?

Ok, that's really weird. Does this only happen if we unwrap the request in `SessionCsrfTokenStrategy`?

Regarding the other changes proposed in the PR: Everything looks fine. Thanks for applying the modification. The only thing missing: I noticed that a few of the newly created files contain Oracle copyright notices. I guess this was copy & pasted from other files. It would be great if you could adjust the headers.

**Comment by erdlet** [comment]
> Ok, that's really weird. Does this only happen if we unwrap the request in `SessionCsrfTokenStrategy`?

No, this was before this change too. I've no clue what the problem is at this point, as it seems to be something inside the server. I'll try to debug it again, but I'm afraid this is no Krazo specific problem.

> Regarding the other changes proposed in the PR: Everything looks fine. Thanks for applying the modification. The only thing missing: I noticed that a few of the newly created files contain Oracle copyright notices. I guess this was copy & pasted from other files. It would be great if you could adjust the headers.

Thanks for checking the headers, I hope I fixed all of them.

**Comment by erdlet** [comment]
I tested this missing header against JSP and Freemarker and recognized, that this only occurs with JSPs. Maybe this is somehow related to the problem in #132?

**Pull request merged**